src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,201 @@
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (index == 0) {
+            return unlinkFirst();
+        } else if (index == size - 1) {
+            return unlinkLast();
+        } else {
+            return unlink(index);
+        }
     }
 
     @Override
     public boolean remove(T object) {
+        int getIndex = getIndex(object);
+        if (getIndex != -1) {
+            remove(getIndex);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T item;
+        private Node<T> next;
+
+        Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Illegal index");
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> node;
+        if (index < (size >> 1)) {
+            node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        } else {
+            node = last;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+        }
+        return node;
+    }
+
+    private int getIndex(T object) {
+        int i = 0;
+        Node<T> temp = first;
+        for (; i < size; i++) {
+            if (object == null) {
+                if (temp.item == null) {
+                    return i;
+                }
+                temp = temp.next;
+            } else {
+                if (object.equals(temp.item)) {
+                    return i;
+                }
+                temp = temp.next;
+            }
+        }
+        return -1;
+    }
+
+    private void linkFirst(T value) {
+        final Node<T> link = first;
+        final Node<T> node = new Node<>(null, value, link);
+        first = node;
+        if (link == null) {
+            last = node;
+        } else {
+            link.prev = node;
+        }
+        size++;
+    }
+
+    private void linkBefore(T value, int index) {
```

Comment:
```suggestion
    private void linkBefore(T value, Node<T> node) {
```

Reply:
I didn't really get why I should modify these two methods(unlink, linkBefore) so that they receive Node as a parameter. In such a scenario I'll need to modify the logic of public methods where these private methods are used.